samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,582 @@
+		.recv_buf_len = sizeof(ctx->http_recv_buffer)
+	};
+
+	ctx->flash_buffer_len = 0;
+	ctx->http_status_code = 0;
+
+	if (ctx->image_size > 0 && ctx->range_start >= (int)ctx->image_size) {
+		return -ENODATA;
+	}
+
+	range_end = ctx->range_start + sizeof(ctx->flash_buffer) - 1;
+	if (ctx->range_start) {
+		range_end = MIN(range_end, ctx->image_size);
+	}
+
+	/* Prepare headers */
+	snprintf(range_header, sizeof(range_header), "Range: bytes=%d-%d\r\n", ctx->range_start,
+		 range_end);
+
+	/* Send request */
+	ret = http_client_req(ctx->sock, &req, 3000, ctx);
+	if (ret < 0) {
+		printf("HTTP request failed: %s, requested range=%d-%d\n", strerror(-ret),
+		       ctx->range_start, range_end);
+		return ret;
+	}
+
+	/* Handle HTTP status codes - centralized error management */
+	if (ctx->http_status_code >= 300) {
+		printf("HTTP error %d, requested range=%d-%d\n", ctx->http_status_code,
+		       ctx->range_start, range_end);
+		return -EPROTO;
+	}
+	return 0;
+}
+
+/*
+ * @brief Process firmware update data packets
+ */
+static int ota_load_firmware(struct app_ctx *ctx)
+{
+	uint32_t ota_calc_image_size = 0;
+	uint32_t status = 0;
+	int buf_offset = 0;
+
+	__ASSERT(ctx->flash_buffer_len > 0, "No firmware data to process");
+
+	if (ctx->range_start == 0) {
+		buf_offset = sizeof(sl_si91x_firmware_header_t);
+		status = sl_wifi_get_firmware_size(ctx->flash_buffer, &ota_calc_image_size);
+		if (status != SL_STATUS_OK) {
+			printf("Unable to fetch firmware size. Status: 0x%08x\n", status);
+			return -EIO;
+		}
+
+		if (ctx->image_size != ota_calc_image_size) {
+			printf("Corrupted payload (image size mismatch: %u bytes != %u bytes)\n",
+			       ctx->image_size, ota_calc_image_size);
+			return -EINVAL;
+		}
+
+		printf("Firmware size: %u bytes\n", ctx->image_size);
+
+		status = sl_si91x_fwup_start(ctx->flash_buffer);
+		if (status != SL_STATUS_OK) {
+			printf("Failed to load firmware header (0x%08x)\n", status);
+			return -EINVAL;
+		}
+	}
+
+	status = sl_si91x_fwup_load(ctx->flash_buffer + buf_offset,
+				    ctx->flash_buffer_len - buf_offset);
+
+	printf("Firmware load success for requested range=%d-%d\n", ctx->range_start,
+	       ctx->range_start + ctx->flash_buffer_len - 1);
+
+	/* Check if firmware update is completed */
+	if (status == SL_STATUS_SI91X_FW_UPDATE_DONE) {
+		printf("Firmware update completed. Rebooting...\n");
+		sys_reboot();
+	}
+	if (status != SL_STATUS_OK) {
+		return -EIO;
+	}
+	return 0;
+}
+
+/*
+ * @brief Main OTA application state machine
+ */
+static int ota_application_start(struct app_ctx *ctx)
+{
+	int ret = 0;
+
+	if (!ctx->connected) {
+		k_event_clear(&ctx->events, WIFI_SCAN_DONE_EVENT | WIFI_SECURITY_TYPE_SET);
+		ota_wifi_start_scan(ctx);
+
+		ret = k_event_wait_safe(&ctx->events, WIFI_SCAN_DONE_EVENT, false, K_SECONDS(10));
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = k_event_wait_safe(&ctx->events, WIFI_SECURITY_TYPE_SET, false, K_SECONDS(1));
+		if (ret < 0) {
+			return ret;
+		}
```

Comment:
Probably you want to wait `WIFI_SCAN_DONE_EVENT` *or* `WIFI_SECURITY_TYPE_SET`:

    ret = k_event_wait_safe(&ctx->events, WIFI_SECURITY_TYPE_SET | WIFI_SCAN_DONE_EVENT,
                            false, K_SECONDS(10));

Also note `k_event_wait_safe()` never return < 0. I believe we don't need to check the return value.